File: platform/chat-service/src/services/s3_service.py
import logging
import os
import tempfile

import boto3
from botocore.exceptions import ClientError

logger = logging.getLogger(__name__)


class S3Service:
    """
    S3 Service with IAM Role (IRSA) authentication for production.
    
    Chat service MUST use its own bucket (chat-service-storage-prod) if needed.
    In production, uses IAM roles (no credentials).
    In development, uses LocalStack with test credentials.
    """
    def __init__(self):
        # CRITICAL: Bucket name MUST be service-specific (chat-service bucket if S3 is used)
        self.bucket_name = os.getenv("S3_BUCKET_NAME")
        if not self.bucket_name:
            raise ValueError(
                "S3_BUCKET_NAME environment variable is required. "
                "Chat service must use 'chat-service-storage-prod' bucket if S3 storage is needed"
            )
        
        self.environment = os.getenv("ENVIRONMENT", "development")
        self.region = os.getenv("AWS_REGION", "us-east-1")
        
        if self.environment == "production":
            # Production: Use IAM role (IRSA), NO credentials
            self.s3_client = boto3.client("s3", region_name=self.region)
            logger.info(
                f"Chat S3 client initialized for PRODUCTION (IRSA) - "
                f"Bucket: {self.bucket_name}, Region: {self.region}"
            )
        else:
            # Development: LocalStack
            self.endpoint_url = os.getenv("S3_ENDPOINT_URL", "http://localstack:4566")
            self.s3_client = boto3.client(
                "s3",
                endpoint_url=self.endpoint_url,
                aws_access_key_id=os.getenv("AWS_ACCESS_KEY_ID", "test"),
                aws_secret_access_key=os.getenv("AWS_SECRET_ACCESS_KEY", "test"),
                region_name=self.region,
            )
            logger.info(
                f"Chat S3 client initialized for DEVELOPMENT (LocalStack) - "
                f"Bucket: {self.bucket_name}, Endpoint: {self.endpoint_url}"
            )

    def upload_file(self, content: bytes, s3_key: str, content_type: str | None = None) -> str:
        """Upload file bytes to THIS service's bucket only (chat-service bucket)"""
        extra_args = {}
        if content_type:
            extra_args["ContentType"] = content_type
        try:
            self.s3_client.put_object(Bucket=self.bucket_name, Key=s3_key, Body=content, **extra_args)
            s3_url = f"s3://{self.bucket_name}/{s3_key}"
            logger.info(f"Uploaded file to {s3_url}")
            return s3_url
        except ClientError as e:
            logger.error(f"Error uploading to S3 bucket '{self.bucket_name}': {e}")
            raise

    def download_file(self, s3_url: str) -> bytes:
        """Download file bytes from S3 URL (must be from THIS service's bucket)"""
        bucket, key = self._parse_s3_url(s3_url)
        
        # Security check: Ensure we're only accessing our own bucket
        if bucket != self.bucket_name:
            error_msg = (
                f"SECURITY VIOLATION: Chat service attempted to access bucket '{bucket}' "
                f"but is only authorized for '{self.bucket_name}'"
            )
            logger.error(error_msg)
            raise PermissionError(error_msg)
        
        try:
            response = self.s3_client.get_object(Bucket=bucket, Key=key)
            return response["Body"].read()
        except ClientError as e:
            logger.error(f"Error downloading from S3: {e}")
            raise

    def download_to_temp_file(self, s3_url: str) -> str:
        """Download to a temp file and return path."""
        content = self.download_file(s3_url)
        with tempfile.NamedTemporaryFile(delete=False, suffix=".tmp") as tmp:
            tmp.write(content)
            temp_path = tmp.name
        logger.info(f"Downloaded to temp file: {temp_path}")
        return temp_path

    def generate_presigned_url(self, s3_key: str, expiration: int = 3600) -> str:
        """Generate presigned GET URL."""
        if s3_key.startswith("s3://"):
            bucket, key = self._parse_s3_url(s3_key)
            # Security check
            if bucket != self.bucket_name:
                raise PermissionError(
                    f"Cannot generate URL for bucket '{bucket}' - only authorized for '{self.bucket_name}'"
                )
        else:
            key = s3_key
            
        try:
            url = self.s3_client.generate_presigned_url(
                "get_object", Params={"Bucket": self.bucket_name, "Key": key}, ExpiresIn=expiration
            )
            # Replace internal Docker hostname with localhost (development only)
            if self.environment != "production" and url and "localstack:" in url:
                url = url.replace("localstack:", "localhost:")
            return url
        except ClientError as e:
            logger.error(f"Error generating presigned URL: {e}")
            return ""

    def _parse_s3_url(self, s3_url: str):
        """Helper to extract bucket and key from s3:// URL"""
        if s3_url.startswith("s3://"):
            parts = s3_url[5:].split("/", 1)
            if len(parts) == 2:
                return parts[0], parts[1]
        raise ValueError(f"Invalid S3 URL format: {s3_url}")


# Global instance - will raise ValueError if S3_BUCKET_NAME not set
s3_service = S3Service()

File: stt-service/src/services/s3.service.js
const { S3Client, PutObjectCommand, GetObjectCommand, DeleteObjectCommand } = require("@aws-sdk/client-s3");
const logger = require("../utils/logger");

/**
 * S3 Service with IAM Role (IRSA) authentication for production.
 * 
 * STT service MUST use its own bucket (stt-service-storage-prod).
 * In production, uses IAM roles (no credentials).
 * In development, uses LocalStack with test credentials.
 */

// CRITICAL: Bucket name MUST be service-specific
const BUCKET_NAME = process.env.S3_BUCKET_NAME;
if (!BUCKET_NAME) {
  throw new Error(
    "S3_BUCKET_NAME environment variable is required. " +
    "STT service must use 'stt-service-storage-prod' bucket"
  );
}

const ENVIRONMENT = process.env.ENVIRONMENT || 'development';
const AWS_REGION = process.env.AWS_REGION || 'us-east-1';

let s3Config;

if (ENVIRONMENT === 'production') {
  // Production: Use IAM role (IRSA), NO credentials
  s3Config = {
    region: AWS_REGION,
  };
  logger.info(`STT S3 client initialized for PRODUCTION (IRSA) - Bucket: ${BUCKET_NAME}, Region: ${AWS_REGION}`);
} else {
  // Development: LocalStack
  const S3_ENDPOINT = process.env.S3_ENDPOINT || 'http://localstack:4566';
  s3Config = {
    region: AWS_REGION,
    endpoint: S3_ENDPOINT,
    credentials: {
      accessKeyId: process.env.AWS_ACCESS_KEY_ID || 'test',
      secretAccessKey: process.env.AWS_SECRET_ACCESS_KEY || 'test',
    },
    forcePathStyle: true, // Required for LocalStack
  };
  logger.info(`STT S3 client initialized for DEVELOPMENT (LocalStack) - Bucket: ${BUCKET_NAME}, Endpoint: ${S3_ENDPOINT}`);
}

const s3Client = new S3Client(s3Config);

/**
 * Upload audio file to S3 (STT service's own bucket only)
 */
const uploadToS3 = async (buffer, filename, contentType) => {
  try {
    const key = filename.startsWith("audio/") ? filename : `audio/${Date.now()}-${filename}`;

    const command = new PutObjectCommand({
      Bucket: BUCKET_NAME,
      Key: key,
      Body: buffer,
      ContentType: contentType,
    });

    await s3Client.send(command);

    const s3Uri = `s3://${BUCKET_NAME}/${key}`;
    logger.info(`Uploaded audio to ${s3Uri}`);
    return s3Uri;
  } catch (error) {
    logger.error(`Error uploading to S3 bucket '${BUCKET_NAME}':`, error);
    throw error;
  }
};

/**
 * Get file from S3 (STT bucket only)
 */
const getFromS3 = async (key) => {
  try {
    const command = new GetObjectCommand({
      Bucket: BUCKET_NAME,
      Key: key,
    });

    const response = await s3Client.send(command);
    const chunks = [];
    for await (const chunk of response.Body) {
      chunks.push(chunk);
    }
    return Buffer.concat(chunks);
  } catch (error) {
    logger.error("S3 get error:", error);
    throw error;
  }
};

/**
 * Delete file from S3
 */
const deleteFromS3 = async (key) => {
  try {
    const command = new DeleteObjectCommand({
      Bucket: BUCKET_NAME,
      Key: key,
    });

    await s3Client.send(command);
    logger.info(`Deleted file from S3: ${key}`);
  } catch (error) {
    logger.error("S3 delete error:", error);
    throw error;
  }
};

/**
 * Download file from S3 URL (must be from STT bucket only)
 */
const downloadFromS3 = async (s3Uri) => {
  try {
    // Parse S3 URI
    let bucket, key;

    if (s3Uri.startsWith("s3://")) {
      const parts = s3Uri.replace("s3://", "").split("/");
      bucket = parts[0];
      key = parts.slice(1).join("/");
    } else {
      // Assume it's just a key
      bucket = BUCKET_NAME;
      key = s3Uri;
    }

    // Security check: Ensure we're only accessing our own bucket
    if (bucket !== BUCKET_NAME) {
      const error = new Error(
        `SECURITY VIOLATION: STT service attempted to access bucket '${bucket}' ` +
        `but is only authorized for '${BUCKET_NAME}'`
      );
      logger.error(error.message);
      throw error;
    }

    const command = new GetObjectCommand({
      Bucket: bucket,
      Key: key,
    });

    const response = await s3Client.send(command);
    const chunks = [];
    for await (const chunk of response.Body) {
      chunks.push(chunk);
    }
    const buffer = Buffer.concat(chunks);

    logger.info(`Downloaded from S3: ${bucket}/${key} (${buffer.length} bytes)`);
    return buffer;
  } catch (error) {
    logger.error("S3 download error:", error);
    throw error;
  }
};

module.exports = { uploadToS3, getFromS3, deleteFromS3, downloadFromS3 };
